mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp

```diff
@@ -1710,18 +1723,31 @@ LogicalResult TargetOp::verify() {
   if (std::distance(teamsOps.begin(), teamsOps.end()) > 1)
     return emitError("target containing multiple teams constructs");
 
-  if (!isTargetSPMDLoop() && getTripCount())
-    return emitError("trip_count set on non-SPMD target region");
-
-  if (teamsOps.empty()) {
-    if (getNumTeamsLower() || getNumTeamsUpper() || getTeamsThreadLimit())
-      return emitError(
-          "num_teams and teams_thread_limit arguments only allowed if there is "
-          "an omp.teams child operation");
-  } else {
-    if (failed(verifyNumTeamsClause(*this, getNumTeamsLower(),
-                                    getNumTeamsUpper())))
-      return failure();
+  // Check that host_eval values are only used in legal ways.
+  bool isTargetSPMD = isTargetSPMDLoop();
+  for (Value hostEvalArg :
+       cast<BlockArgOpenMPOpInterface>(getOperation()).getHostEvalBlockArgs()) {
+    for (Operation *user : hostEvalArg.getUsers()) {
+      if (auto teamsOp = dyn_cast<TeamsOp>(user)) {
+        if (llvm::is_contained({teamsOp.getNumTeamsLower(),
+                                teamsOp.getNumTeamsUpper(),
+                                teamsOp.getThreadLimit()},
+                               hostEvalArg))
+          continue;
+      } else if (auto parallelOp = dyn_cast<ParallelOp>(user)) {
+        if (isTargetSPMD && hostEvalArg == parallelOp.getNumThreads())
+          continue;
+      } else if (auto loopNestOp = dyn_cast<LoopNestOp>(user)) {
+        if (isTargetSPMD &&
+            (llvm::is_contained(loopNestOp.getLoopLowerBounds(), hostEvalArg) ||
+             llvm::is_contained(loopNestOp.getLoopUpperBounds(), hostEvalArg) ||
+             llvm::is_contained(loopNestOp.getLoopSteps(), hostEvalArg))) {
+          continue;
+        }
+      }
+      return emitOpError() << "host_eval argument illegal use in '"
```

Comment:
This is a nit, so feel free to ignore, but we could have higher granularity in terms of this error message. For instance, we could differentiate between illegal operations using `hostEvalArg` and legal operations using it on illegal clauses 

```
if(llvm::is_contained({teamOp.getNumTeamsLower(), teamsOp.getNumTeamsUpper(), teamsOp.getThreadLimit()}, hostEvalArg))
     continue;
else
     return emitOpError() << "host_eval argument used on illegal clause in " << user->getName() << "operation";
```